src/main/java/org/opensearch/geospatial/index/query/xypoint/XYPointQueryVisitor.java

```diff
@@ -0,0 +1,210 @@
+import org.opensearch.geometry.LinearRing;
+import org.opensearch.geometry.MultiLine;
+import org.opensearch.geometry.MultiPoint;
+import org.opensearch.geometry.MultiPolygon;
+import org.opensearch.geometry.Point;
+import org.opensearch.geometry.Polygon;
+import org.opensearch.geometry.Rectangle;
+import org.opensearch.geometry.ShapeType;
+import org.opensearch.index.mapper.MappedFieldType;
+import org.opensearch.index.query.QueryShardContext;
+import org.opensearch.index.query.QueryShardException;
+
+/**
+ * Geometry Visitor to create a query to find all cartesian XYPoints
+ * that comply ShapeRelation with all other XYShapes objects
+ */
+@AllArgsConstructor
+public class XYPointQueryVisitor implements GeometryVisitor<Query, RuntimeException> {
+    QueryShardContext context;
+    MappedFieldType fieldType;
+    String fieldName;
+
+    /**
+     * @param line  input geometry {@link Line}
+     * @return {@link Query} instance from XYPointField.XYPointInGeometryQuery
+     * throws QueryShardException
+     */
+    @Override
+    public Query visit(Line line) {
+        throw new QueryShardException(
+            context,
+            String.format(Locale.getDefault(), "Field [%s] found an unsupported shape [%s]", fieldName, ShapeType.LINESTRING.name())
+        );
+    }
+
+    /**
+     * @param ring  input geometry {@link LinearRing}
+     * @return {@link Query} instance from XYPointField.XYPointInGeometryQuery
+     * throws QueryShardException
+     */
+    @Override
+    public Query visit(LinearRing ring) {
+        throw new QueryShardException(
+            context,
+            String.format(Locale.getDefault(), "Field [%s] found an unsupported shape [%s]", fieldName, ShapeType.LINEARRING.name())
+        );
+    }
+
+    /**
+     * @param multiLine  input geometry {@link MultiLine}
+     * @return {@link Query} instance from XYPointField.XYPointInGeometryQuery
+     * throws QueryShardException
+     */
+    @Override
+    public Query visit(MultiLine multiLine) {
+        throw new QueryShardException(
+            context,
+            String.format(Locale.getDefault(), "Field [%s] found an unsupported shape [%s]", fieldName, ShapeType.MULTILINESTRING)
+        );
+    }
+
+    /**
+     * @param multiPoint  input geometry {@link MultiPoint}
+     * @return {@link Query} instance from XYPointField.XYPointInGeometryQuery
+     * throws QueryShardException
+     */
+    @Override
+    public Query visit(MultiPoint multiPoint) {
+        throw new QueryShardException(
+            context,
+            String.format(Locale.getDefault(), "Field [%s] found an unsupported shape [%s]", fieldName, ShapeType.MULTIPOINT)
+        );
+    }
+
+    /**
+     * @param point  input geometry {@link Point}
+     * @return {@link Query} instance from XYPointField.XYPointInGeometryQuery
+     * throws QueryShardException
+     */
+    @Override
+    public Query visit(Point point) {
+        throw new QueryShardException(
```

Comment:
nit: Shall we create a private method to build and raise this exception instead of rewriting multiple times?

Reply:
updated